src/qps/impls/smalxe/smalxe.c

```diff
@@ -616,19 +616,22 @@ PETSC_STATIC_INLINE PetscErrorCode QPSConverged_Inner_SMALXE_Monitor_Inner(QPS q
 
 #undef __FUNCT__
 #define __FUNCT__ "QPSConverged_Inner_SMALXE"
-PetscErrorCode QPSConverged_Inner_SMALXE(QPS qps_inner,QP qp_inner,PetscInt i,PetscReal gnorm,KSPConvergedReason *reason,void *ctx)
+PetscErrorCode QPSConverged_Inner_SMALXE(QPS qps_inner,KSPConvergedReason *reason,void *ctx)
 {
   QPSConvergedCtx_Inner_SMALXE *cctx = (QPSConvergedCtx_Inner_SMALXE*) ctx;
   QPS qps_outer = cctx->qps_outer;
   QPS_SMALXE *smalxe = (QPS_SMALXE*)qps_outer->data;
-  QP qp_outer = cctx->qp_outer;
+  QP qp_inner = qps_inner->solQP;
   Vec u = qp_inner->x;
+  PetscInt i = qps_inner->iteration;
+  PetscReal gnorm = qps_inner->rnorm;
   MPI_Comm comm;
   
   PetscFunctionBegin;
-  FLLOP_ASSERT(qp_inner->parent == qp_outer,"qp_inner->parent == qp_outer");
-  FLLOP_ASSERT(qp_inner->x == qp_outer->x, "qp_inner->x == qp_outer->x");
+  FLLOP_ASSERT(qp_inner->parent == cctx->qp_outer,"qp_inner->parent == qp_outer");
+  FLLOP_ASSERT(qp_inner->x == cctx->qp_outer->x, "qp_inner->x == qp_outer->x");
   TRY( PetscObjectGetComm((PetscObject)qps_inner,&comm) );
+  if (!qps_inner->setupcalled) FLLOP_SETERRQ(comm, PETSC_ERR_ARG_WRONGSTATE, "Inner QPSSetUp() not yet called");
```

Comment:
I do not think that this is neccessary.

Reply:
but does not hurt neither - kind of assertion check

Reply:
My point is that we know that QPSSetUp(qps_inner) was called because at this point QPSSetUp(qps_outter) was called and created qps_inner.

Reply:
ok, I moved all the checks here to Solve and only in debug mode